### PR TITLE
[nest] Add support for Smart Device Management (SDM) API

### LINKED ISSUE #8664: [nest] Add support for Smart Device Management (SDM) API
The current API used by the binding is based on the Works With Nest (WWN) API for which new users can no longer sign up (see [What's happening at Nest?](https://nest.com/whats-happening/)).

Recently a new Smart Device Management API has become available which again allows for controlling Nest devices, see:
* [Device Access](https://developers.google.com/nest/device-access/)
* [Smart Device Management (SDM) API](https://developers.google.com/nest/device-access/api/)

There is a non-refundable $5 registration fee for creating a developer account to gain access to this API.

See also the following community topics:
* [Google Nest Device Access Console now available](https://community.openhab.org/t/google-nest-device-access-console-now-available/105404?u=wborn)
* [Works With Nest - Deprecating on August 31, 2019](https://community.openhab.org/t/works-with-nest-deprecating-on-august-31-2019/73974?u=wborn)

I'll start implementing this SDM API in the existing binding so existing WWN API users can migrate over time.
The new SDM API may also not provide the same functionality.
Most likely the WWN API will be retired and can then be removed from the binding.

### PR DESCRIPTION
**Pull request by wborn**
* Reworks WWN implementation so that the thing types have a wwn_ prefix and the classes have a WWN prefix and reside in a 'wwn' package
* Adds an SDM implementation which is also based on: https://github.com/bhigg-code/openhab-addons/tree/2.5.x/bundles/org.openhab.binding.nestdeviceaccess
* Adds unit tests for (de)serialization of the SDM and Pub/Sub API requests and responses
* Updates the binding documentation for the changes and additions

Fixes #8664

---

The following bundles can be used for testing the changes:

* [org.openhab.binding.nest-2.5.13.20210530.jar](https://github.com/wborn/openhab-addons/releases/download/nest-sdm-api-20210530/org.openhab.binding.nest-2.5.13.20210530.jar) (retro OH versions: 2.5.0, 2.5.1, ... 2.5.10, 2.5.11, 2.5.12)
* [org.openhab.binding.nest-3.0.3.20210530.jar](https://github.com/wborn/openhab-addons/releases/download/nest-sdm-api-20210530/org.openhab.binding.nest-3.0.3.20210530.jar) (older OH versions: 3.0.0, 3.0.1, 3.0.2, 3.1.0.M1, 3.1.0.M2, 3.1.0.M3, 3.1.0.M4)
* [org.openhab.binding.nest-3.1.0.20210530.jar](https://github.com/wborn/openhab-addons/releases/download/nest-sdm-api-20210530/org.openhab.binding.nest-3.1.0.20210530.jar) (newer OH versions: 3.1.0-SNAPSHOT, 3.1.0.M5)

For documentation, see the updated [README.md](https://github.com/wborn/openhab-addons/blob/ee410fe865fe9bd0759abe774acf7d944e8477ff/bundles/org.openhab.binding.nest/README.md#nest-binding).

### CODE REVIEW

**Comment by junalmeida** [comment]
@wborn Suggestion: 

In `sdm-config.xml` give brief instructions how the user can create the "Device Access Console" project, and "Google Cloud Console" oauth credentials. I had no idea (and I couldn't find anywhere) whether to create a web or desktop oauth type. 
And after figuring out that I need a desktop type, the redirect-url from google instructions is a mistake, as there is no google.com uri allowed for a desktop oauth type. 

The correct url to get the auth code is: 
https://nestservices.google.com/partnerconnections/{Project_ID}/auth?redirect_uri=urn:ietf:wg:oauth:2.0:oob&access_type=offline&prompt=consent&client_id={Client_ID}&response_type=code&scope=https://www.googleapis.com/auth/sdm.service

With `redirect_uri` set to `urn:ietf:wg:oauth:2.0:oob`. I could only find this out by reading the code. 


Also, I forgot to enable the correct API, so it would be nice to see that instruction in the config page as well. If API is not enabled, discovery will fail silently because the resultant json is: 
```
"error": {
    "code": 403,
    "message": "Smart Device Management API has not been used in project XXXXXX before or it is disabled. Enable it by visiting https://console.developers.google.com/apis/api/smartdevicemanagement.googleapis.com/overview?project=XXXXXXX then retry. If you enabled this API recently, wait a few minutes for the action to propagate to our systems and retry.",
``` 

After passing by all these, I can now successfully connect to my Nest thermostat as the old API is not available for me. I'm using your latest code from the source of this PR compiled locally. 

I don't know yet how to enable Pub/Sub because I can only see the Pub/Sub topic id. Not sure how to fill it out.

Thanks for your awesome work and I hope this can get merged to openhab soon.

**Comment by wborn** [comment]
Yes adding some docs on how to configure everything is one of the reasons why this PR is still WIP. 🙂 Getting authorization code for accessing Pub/Sub is very similar to how it is done with [Google Cloud TTS](https://www.openhab.org/addons/voice/googletts/#obtaining-credentials). The configuration is a lot easier now it can all be done by creating an account thing using the UI and configuring the authorization code to get the access/refresh tokens.

**Comment by junalmeida** [comment]
Hi, just to let you know, I pulled your changes, it compiled successfully, but after I pushed the jar to my server (3.1.0M4), I got the following: 
```
2021-05-29 10:36:48.110 [WARN ] [org.apache.felix.fileinstall        ] - Error while starting bundle: file:/usr/share/openhab/addons/org.openhab.binding.nest-3.1.0-SNAPSHOT.jar
org.osgi.framework.BundleException: Could not resolve module: org.openhab.binding.nest [300]
  Unresolved requirement: Import-Package: javax.measure; version="[2.1.0,3.0.0)"
	at org.eclipse.osgi.container.Module.start(Module.java:463) ~[org.eclipse.osgi-3.16.200.jar:?]
	at org.eclipse.osgi.internal.framework.EquinoxBundle.start(EquinoxBundle.java:440) ~[org.eclipse.osgi-3.16.200.jar:?]
	at org.apache.felix.fileinstall.internal.DirectoryWatcher.startBundle(DirectoryWatcher.java:1260) [bundleFile:3.6.8]
	at org.apache.felix.fileinstall.internal.DirectoryWatcher.startBundles(DirectoryWatcher.java:1233) [bundleFile:3.6.8]
	at org.apache.felix.fileinstall.internal.DirectoryWatcher.doProcess(DirectoryWatcher.java:520) [bundleFile:3.6.8]
	at org.apache.felix.fileinstall.internal.DirectoryWatcher.process(DirectoryWatcher.java:365) [bundleFile:3.6.8]
	at org.apache.felix.fileinstall.internal.DirectoryWatcher.run(DirectoryWatcher.java:316) [bundleFile:3.6.8]
```
Not sure if this is something missing on my end, but I could only get past this problem by adding the following to `pom.xml`:
```
  <dependencies>
    <dependency>
      <groupId>javax.measure</groupId>
      <artifactId>unit-api</artifactId>
      <version>1.0</version>
    </dependency>
  </dependencies>
````
If this is not the correct way to solve it let me know.

**Comment by wborn** [comment]
This PR is now ready for review @openhab/add-ons-maintainers. Hopefully you can find some time to review it and it can be merged before OH 3.1 is released. :slightly_smiling_face:

There are links to JARs in the [initial PR comment](https://github.com/openhab/openhab-addons/pull/8947#issue-514407686) which can be used to test the changes of this PR with all OH 3 versions.

Because of some dependency changes, the `-Dohc.version=3.0.2` option needs to be added to Maven build commands to generate a bundle that has compatible dependencies with older OH 3.x releases @junalmeida.

**Comment by techie66** [comment]
I could only get an authorization code after using response_type=code in the URI. Without it, the process would fail to complete with no error.

May I also propose generating the proper link to get this code from the settings page?

File: bundles/org.openhab.binding.nest/src/main/java/org/openhab/binding/nest/internal/sdm/handler/SDMAccountHandler.java [hunk 1]
```diff
@@ -0,0 +1,332 @@
...
+        listeners.remove(deviceId, listener);
+    }
+
+    public @Nullable SDMAPI getAPI() {
+        return sdmAPI;
+    }
+
+    private void handlePubSubMessage(PubSubMessage message) {
+        String messageId = message.messageId;
+        String json = new String(Base64.getDecoder().decode(message.data));
```
Comment (fwolter) [inline bundles/org.openhab.binding.nest/src/main/java/org/openhab/binding/nest/internal/sdm/handler/SDMAccountHandler.java:284]: Can you specify the encoding you expect? E.g. StandardCharsets.UTF_8

**Review by fwolter — CHANGES_REQUESTED** [review]

**Comment by fwolter** [comment]
Should the "not backward compatible" label be set?

**Comment by junalmeida** [comment]
Generating the correct link on settings page after filling out the project id and client id would be awesome IMO

**Comment by wborn** [comment]
I've fixed and improved the docs @techie66. That part got lost somewhere while copy/pasting. :+1:

**Comment by wborn** [comment]
Hey I've also backported all the Nest SDM changes of my PR to 2.5.x for all the retro openHAB-ers. :slightly_smiling_face: 
I've added the JAR to the [initial PR comment](https://github.com/openhab/openhab-addons/pull/8947#issue-514407686). It seems to work equally well from the brief testing I gave it.

File: bundles/org.openhab.binding.nest/src/main/java/org/openhab/binding/nest/internal/sdm/handler/SDMAccountHandler.java (hunk 1 above)
Comment (wborn) [reply bundles/org.openhab.binding.nest/src/main/java/org/openhab/binding/nest/internal/sdm/handler/SDMAccountHandler.java]: Yes that's better. You never know what might happen on Windows. :-)

**Review by fwolter — APPROVED** [review]

**Comment by wborn** [comment]
It's nice to see the PR made it into OH 3.1! :partying_face::balloon:

**Comment by resetpointer** [comment]
is there a reason why OATH2 is implemented as a "web application" instead of a "desktop application".  Openhab is certainly is not a web-application, it would make more sense if the OATH2 account is registered as a "desktop application".

Is it possible to update the nest_sdm addon to support OATH2 account of type "desktop application" as well.  thanks.

**Comment by wborn** [comment]
> is there a reason why OATH2 is implemented as a "web application" instead of a "desktop application". Openhab is certainly is not a web-application, it would make more sense if the OATH2 account is registered as a "desktop application".

See: https://support.google.com/cloud/answer/6158849#zippy=%2Cweb-applications%2Cnative-applications

> A web application is accessed by web browsers over a network.
>
> * Applications that use client-side JavaScript to access Google APIs must specify authorized JavaScript origins. The origins identify the domains from which your application can send API requests.
> * Applications that access Google APIs from a server (often using languages and frameworks like Node.js, Java, .NET, and Python) must specify authorized redirect URIs. The redirect URIs are the endpoints of your application server to which the OAuth 2.0 server can send responses.

It would be nice if when we can use openHAB in the redirect URI so it is easier to configure this.

**Comment by resetpointer** [comment]
Wborn,

I reviewed the code base of TTS and nest-sdm yesterday.  What I observed was that the application type used to be "tv and media devices", which is similar to desktop application type.  Since Google OATH2 depreciated OOB (out of bound redirect), openhab began to change the application type to web application, and I believe this was recommended by the SDM api.

Regardless, the authentication flow is too complicated for a typical end user.  This whole process can be simplified by implementing a redirect backend, that can be hosted on openhab.org or someone in the cloud that everyone can use.  This whole process of issuing manual post request and then copy/past authorization code is just too tedious, we need to simplify the common OATH2 verification process for OH users.

Thanks for the prompt reply, btw.

**Comment by wborn** [comment]
> This whole process can be simplified by implementing a redirect backend, that can be hosted on openhab.org or somewhere in the cloud that everyone can use.

I think it should be possible to redirect the browser to your own OH instance even if it does not expose any ports to the Internet. Some bindings already have a servlet for handling such a redirect (see [SpotifyAuthServlet](https://github.com/wborn/openhab-addons/blob/main/bundles/org.openhab.binding.spotify/src/main/java/org/openhab/binding/spotify/internal/SpotifyAuthServlet.java)). But I think it would be better if such flows are also supported by the openHAB Core so servlets handling the redirect do not have to be duplicated in each binding that uses (Google) OAuth 2. :slightly_smiling_face:

**Comment by resetpointer** [comment]
> > This whole process can be simplified by implementing a redirect backend, that can be hosted on openhab.org or somewhere in the cloud that everyone can use.
> 
> I think it should be possible to redirect the browser to your own OH instance even if it does not expose any ports to the Internet. Some bindings already have a servlet for handling such a redirect (see [SpotifyAuthServlet](https://github.com/wborn/openhab-addons/blob/main/bundles/org.openhab.binding.spotify/src/main/java/org/openhab/binding/spotify/internal/SpotifyAuthServlet.java)). But I think it would be better if such flows are also supported by the openHAB Core so servlets handling the redirect do not have to be duplicated in each binding that uses (Google) OAuth 2. slightly_smiling_face


good idea!  Please address it to the core developers.

**Pull request merged**